bentoml/service.py

```diff
@@ -421,30 +437,69 @@ def env(self):
 
     @classmethod
     def name(cls):  # pylint:disable=method-hidden
-        if cls._bento_service_name is not None and isidentifier(
-            cls._bento_service_name
-        ):
+        if cls._bento_service_name is not None:
+            if not isidentifier(cls._bento_service_name):
+                raise ValueError(
+                    'BentoService#_bento_service_name must be valid python identifier'
+                    'matching regex `(letter|"_")(letter|digit|"_")*`'
+                )
+
             return cls._bento_service_name
         else:
             # Use python class name as service name
             return cls.__name__
 
+    def set_version(self, version_str=None):
+        if version_str is None:
+            version_str = self._new_version_str()
+
+        if self._version_major is not None and self._version_minor is not None:
+            # BentoML uses semantic versioning for BentoService distribution
+            # when user specified the MAJOR and MINOR version number along with
+            # the BentoService class definition with '@ver' decorator.
+            # The parameter version(or auto generated version) here will be used as
+            # PATCH field in the final version:
+            version_str = ".".join(
+                [str(self._version_major), str(self._version_minor), version_str]
+            )
+
+        _validate_version_str(version_str)
+        self._bento_service_version = version_str
+        return self._bento_service_version
+
+    def _new_version_str(self):
+        """
+        Default new version string generator function
+        User can override this function in their custom BentoService to get a customized
+        version format
+        """
+        time_obj = datetime.now()
+        date_string = time_obj.strftime("%Y_%m_%d")
+        random_hash = uuid.uuid4().hex[:8]
+
+        return date_string + "_" + random_hash
+
     @property
     def version(self):
-        try:
-            return self.__class__._bento_service_version
-        except AttributeError:
-            raise BentoMLException(
-                "Only BentoService loaded from archive has version attribute"
-            )
+        if self._bento_service_version is None:
+            self.set_version(self._new_version_str())
+
+        return self._bento_service_version
 
     def save(self, base_path=None, version=None):
         track_save(self)
-        if base_path:
-            repo = LocalRepository(base_path)
-        else:
-            repo = get_default_repository()
-        return repo.add(self, version=version)
+        if version is not None:
+            self.set_version(version)
+
+        # TODO: Callding BentoRepository directly for now, this should be changed to
```

Comment:
will leave the Bento related RPC implementation to follow up PR, those will be including writing Bento metadata to DB and support listing saved Bentos